src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -215,30 +247,51 @@ void WatchHierarchy(IVsHierarchy hierarchyToWatch)
                 }
 
                 // We may have multiple projects with the same hierarchy, but we can use __VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext to distinguish
+                VisualStudioProject? project = null;
                 if (ErrorHandler.Succeeded(hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext, out var contextProjectNameObject)))
                 {
                     WatchHierarchy(hierarchy);
 
                     if (contextProjectNameObject is string contextProjectName)
                     {
-                        var project = _workspace.GetProjectWithHierarchyAndName_NoLock(hierarchy, contextProjectName);
-
-                        if (project != null && projectIds.Contains(project.Id))
-                        {
-                            return project.Id;
-                        }
+                        project = _workspace.GetProjectWithHierarchyAndName(hierarchy, contextProjectName);
                     }
                 }
 
-                // At this point, we should hopefully have only one project that maches by hierarchy. If there's multiple, at this point we can't figure anything
-                // out better.
-                var matchingProjectId = projectIds.FirstOrDefault(id => projectToHierarchyMap.GetValueOrDefault(id, null) == hierarchy);
+                return (hierarchy, project, projectToHierarchyMap);
 
-                if (matchingProjectId != null)
+                void WatchHierarchy(IVsHierarchy hierarchyToWatch)
                 {
-                    return matchingProjectId;
+                    _watchedHierarchiesForDocumentMoniker.Add(moniker, _hierarchyEventSinkCache.GetOrCreate(hierarchyToWatch, static (h, self) => new HierarchyEventSink(h, self), this));
+                }
+            }
+
+            private static ProjectId? GetActiveProjectId(
+                IVsHierarchy? hierarchy,
+                VisualStudioProject? project,
+                ImmutableDictionary<ProjectId, IVsHierarchy?>? projectToHierarchyMap,
+                ImmutableArray<ProjectId> projectIds)
+            {
+                // Keep this method in sync with GetActiveContextProjectIdAndWatchHierarchies.
+
+                if (hierarchy == null)
+                {
+                    // Any item in the RDT should have a hierarchy associated; in this case we don't so there's
+                    // absolutely nothing we can do at this point.
+                    return projectIds.First();
                 }
```

Comment:
i hate this logic.  when do we not have a hierarchy?  should we just blowup instead?